FSharp/Telegram.Bot.Examples.Polling/Program.fs

```diff
@@ -0,0 +1,254 @@
+                        | _                -> usage
+
+                    do! fn
+            }
+
+        let unknownUpdateHandlerAsync (update:Update) = async {
+            Console.WriteLine($"Unknown update type: {update.Type}");
+        }
+
+        let handleUpdateAsync bot (update:Update) cts = async {
+            try
+                let fn =
+                    match update.Type with
+                    | UpdateType.Message            -> botOnMessageReceived update.Message
+                    | UpdateType.EditedMessage      -> botOnMessageReceived update.EditedMessage
+                    | UpdateType.CallbackQuery      -> botOnCallbackQueryReceived update.CallbackQuery
+                    | UpdateType.InlineQuery        -> botOnInlineQueryReceived update.InlineQuery
+                    | UpdateType.ChosenInlineResult -> botOnChosenInlineResultReceived update.ChosenInlineResult
+                    | _                             -> unknownUpdateHandlerAsync update
+                fn |> Async.Start
+            with
+                | ex -> do! handleErrorAsync bot ex cts
+        }
+
+        async {
+            let! me = botClient.GetMeAsync() |> Async.AwaitTask
+            printfn $"Hello, World! I am user {me.Id} and my name is {me.FirstName}."
+            printfn $"Start listening for {me.Username}..."
+
+            use cts = new CancellationTokenSource();
+
+            // There is quite some bit of jugglery here, so requires some explanation:
+            // DefaultUpdateHandler() requires two arguments, both of type Func<_,_,_,_>, and both
+            // of which return a Task. Now, in order to be in the F# domain, we would like to
+            // have this Func<> defined as an F# function: and so we need to explicitely construct
+            // Func<>s with by passing them an inner lambda function.
+            // Now, the inner lambda function needs to return a Task, but we want to use F# async.
+            // We therefore use a Async.StartAsTask to start an async computation and get back a Task.
+            // The last ":>" is for upcasting Task<unit> (returned by the async computations) to their
+            // base class Task to avoid and error that says "the expression expects a Task but we have a
+            // Task<unit> here.". Since F# can already infer the base-type, we simply upcast to "_".
+            // The good thing about doing all this is that handleUpdateAsync and handleErrorAsync are both
+            // in the F# domain - so now can take all advantages of F#!
+            botClient.StartReceiving(DefaultUpdateHandler(
+                                        Func<_,_,_,_>(fun b u t ->  Async.StartAsTask (handleUpdateAsync b u t) :> _),
+                                        Func<_,_,_,_>(fun b e t ->  Async.StartAsTask (handleErrorAsync b e t) :> _)),
+                                        cts.Token)
+        } |> Async.Start
```

Comment:
Can something be made from our side to make this look nicer?

Reply:
Thank you @tuscen for the review!

The problem, as explained in the comments above is that the arguments to `DefaultUpdateHandler` are both `Func<>`s. F# cannot directly handle `Func<>`s, but they can be used from other .NET languages / .NET libraries as F# allows interoperability with them. 

In this specific case, unfortunately (but good for the original use-case in C# code), the function `DefaultUpdateHandler`'s two `Func<>`s return a `Task`. Now, that is not an easy thing - because F# has better mechanisms for handling asynchronous code - and these are abstracted over the underlying Tasking systems: an F# async gives me exception handling, and fault-code-paths for free, while the underlying Task needs special handling. I therefore switched to the F# domain as mentioned above.

An alternative would have been to use the output of `Func<>`s as `Task` itself, and, define all F# functions seen above also using `Task` instead of the abstracted `async` - but our example then would be very specific. In our specific case, all F# asynchronous functions are very simple - so, ideally, a `Task` would not have made much of a difference - but think of a case where these functions could (in the real-world out there, when this example is used as a base) could talk with other systems (such as databases, etc.) where F# `async` is a real boon. That was the reason for switching quickly to F# right at the beginning of the function flow.

Do you still think there should be some change here or my explanation is now in order for you?

Reply:
Thank you for the explanation. I'm not sure I understood everything since I don't know F#, but I think I've got the gist of the explanation.

Everything is fine then. We'll merge your PR later. Thank you for the contribution.

Reply:
THank you very much @tuscen!